File: typr-tester-typr-java/src/java/adventureworks/WithConnection.java
package adventureworks;

import java.sql.Connection;
import java.sql.DriverManager;
import java.sql.SQLException;
import java.util.function.Consumer;
import java.util.function.Function;

public class WithConnection {
  private static final String JDBC_URL =
      "jdbc:postgresql://localhost:6432/Adventureworks?user=postgres&password=password";

  public static <T> T apply(Function<Connection, T> f) {
    try (Connection conn = DriverManager.getConnection(JDBC_URL)) {
      conn.setAutoCommit(false);
      try {
        return f.apply(conn);
      } finally {
        conn.rollback();
      }
    } catch (SQLException e) {
      throw new RuntimeException(e);
    }
  }

  public static void run(Consumer<Connection> f) {
    try (Connection conn = DriverManager.getConnection(JDBC_URL)) {
      conn.setAutoCommit(false);
      try {
        f.accept(conn);
      } finally {
        conn.rollback();
      }
    } catch (SQLException e) {
      throw new RuntimeException(e);
    }
  }
}

File: typr-tester-typr-kotlin/src/test/kotlin/adventureworks/WithConnection.kt
package adventureworks

import java.sql.Connection
import java.sql.DriverManager

object WithConnection {
    private const val JDBC_URL = "jdbc:postgresql://localhost:6432/Adventureworks?user=postgres&password=password"

    fun <T> apply(f: (Connection) -> T): T {
        DriverManager.getConnection(JDBC_URL).use { conn ->
            conn.autoCommit = false
            try {
                return f(conn)
            } finally {
                conn.rollback()
            }
        }
    }

    fun run(f: (Connection) -> Unit) {
        DriverManager.getConnection(JDBC_URL).use { conn ->
            conn.autoCommit = false
            try {
                f(conn)
            } finally {
                conn.rollback()
            }
        }
    }
}
